app/services/course/material/zip_download_service.rb

```diff
@@ -5,8 +5,51 @@ class << self
     # @param [Course::Material::Folder] folder The folder containing the materials.
     # @param [Array<Course::Material>] materials The materials to be downloaded.
     # @return [String] The path to the zip file.
-    def download_and_zip(_folder, _materials)
-      # TODO: To implement
+    def download_and_zip(folder, materials)
+      service = Course::Material::ZipDownloadService.new(folder, materials)
+      service.download_to_base_dir
+      service.zip_base_dir
     end
   end
+
+  # Downloads the materials to the the base directory.
+  def download_to_base_dir
+    materials.each do |material|
+      download_material(material, folder, base_dir)
+    end
+  end
+
+  # Zip the directory and write to the file.
+  #
+  # @return [String] The path to the zip file.
+  def zip_base_dir
+    output_file = base_dir + '.zip'
+    Zip::File.open(output_file, Zip::File::CREATE) do |zip_file|
+      Dir["#{base_dir}/**/**"].each do |file|
+        zip_file.add(file.sub(File.join(base_dir + '/'), ''), file)
+      end
+    end
+
+    output_file
+  end
+
+  private
+
+  attr_accessor :materials, :folder, :base_dir
```

Comment:
Sorry, what do you mean by "does not follow " ?

All these 3 accessors are private already I think ?